### PR TITLE
test: adding testing trigger for running our e2e suite against the NodeAutoProvisioning Preview

### LINKED ISSUE #195: Running the e2e suite we have developed against the NodeAutoProvisioning Preview feature
### Tell us about your request

We should be able to run our e2e tests, against the currently released version of Node Auto Provisioning in any specified region to validate behaviors have rolled out to our users in specific regions. We should develop some mannerism for triggering our standard /test/suites against the managed version of karpenter. 

### Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?

The challenge here will be reworking the testing framework to support BeforeEach and AfterEach validation when running karpenter in different environments. 

Karpenter self hosted doesn't have any crds by default, whereas nap requires customers use an un-weighted default and system-surge crd. We can simply create nodepools with weights to easily sidestep around this problem but the framework needs to account for the presence of these required nap CRDs.

### Are you currently working around this issue?

there is no workaround.

### Additional Context

_No response_

### Attachments

_No response_

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### PR DESCRIPTION
**Pull request by Bryce-Soghigian**
Fixes #195 

**Description**
This PR adds additional functionality to our testing toolkit to run the e2e suites on node auto provisioning, alongside the consolidation e2e suite which only works against node auto provisioning and not fully properly with unmanaged karpenter. 

**How was this change tested?**
- ran the suites on nap
See Runs: 
https://github.com/Azure/karpenter-provider-azure/actions/runs/8193216788/job/22475958956
https://github.com/Azure/karpenter-provider-azure/actions/runs/8150047908/job/22275681008

Then runs validating the standard e2es functionality haven't been tarnished: https://github.com/Azure/karpenter-provider-azure/actions/runs/8232044776
**Does this change impact docs?**
- [ ] Yes, PR includes docs updates
- [ ] Yes, issue opened: # 
- [x] No

**Release Note**

```release-note

```

### CODE REVIEW

**Review by Bryce-Soghigian — COMMENTED** [review]
/test-nap

**Comment by Bryce-Soghigian** [comment]
https://github.com/Azure/karpenter-provider-azure/actions/runs/8149980061
https://github.com/Azure/karpenter-provider-azure/actions/runs/8150047908/job/22275681008

**Review by Bryce-Soghigian — COMMENTED** [review]
/test-nap

**Review by Bryce-Soghigian — COMMENTED** [review]
/nap

**Review by Bryce-Soghigian — COMMENTED** [review]
/nap

**Review by Bryce-Soghigian — COMMENTED** [review]
/test

**Review by Bryce-Soghigian — COMMENTED** [review]
/nap

**Comment by Bryce-Soghigian** [comment]
See the tests pass for nap https://github.com/Azure/karpenter-provider-azure/actions/runs/8193216788/job/22475958956

**Review by Bryce-Soghigian — COMMENTED** [review]
/test

**Comment by Bryce-Soghigian** [comment]
Validating standard tests wtill work and standard workflow triggers still work

**Comment by Bryce-Soghigian** [comment]
running bash hack/boilerplate.sh is not generating the right configuration for the consolidation header for some reason

File: test/suites/consolidation/suite_test.go [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+var _ = AfterEach(func() { env.AfterEach() })
+
+var _ = Describe("Consolidation", func() {
+	It("should consolidate nodes (delete)", Label(debug.NoWatch), Label(debug.NoEvents), func() {
+		provider := env.DefaultAKSNodeClass()
+		nodePool := env.DefaultNodePool(provider)
+		nodePool.Spec.Disruption.ConsolidationPolicy = corev1beta1.ConsolidationPolicyWhenUnderutilized
+		nodePool.Spec.Disruption.ConsolidateAfter = &corev1beta1.NillableDuration{}
+
+		var numPods int32 = 75
```
Comment (charliedmcb) [inline test/suites/consolidation/suite_test.go:58]: We've been typically trying to keep numbers lowers for tests atm. I think 75 pods requiring 1 cpu is far beyond what we currently want for testing. I'd suggest something < 20.

File: test/suites/consolidation/suite_test.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+		})
+
+		selector := labels.SelectorFromSet(dep.Spec.Selector.MatchLabels)
+		env.ExpectCreatedNodeCount("==", 0)
+		env.ExpectCreated(nodePool, provider, dep)
+
+		env.EventuallyExpectHealthyPodCount(selector, int(numPods))
+
+		// reduce the number of pods by 60%
+		dep.Spec.Replicas = to.Ptr[int32](30)
```
Comment (charliedmcb) [inline test/suites/consolidation/suite_test.go:78]: if you want it to be a percentage maybe do a math calculation based on the first number. 

If the first one is 20, than you could have this one be (original * 0.4)

File: test/suites/consolidation/suite_test.go (hunk 1 above)
Comment (charliedmcb) [reply test/suites/consolidation/suite_test.go:58]: My req would actually be to use something closer to 10

File: test/suites/consolidation/suite_test.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+		env.EventuallyExpectHealthyPodCount(selector, int(numPods))
+
+		// 3 nodes due to the anti-affinity rules
+		env.ExpectCreatedNodeCount("==", 3)
+
+		// scaling down the large deployment leaves only small pods on each node
+		largeDep.Spec.Replicas = to.Ptr[int32](0)
+		env.ExpectUpdated(largeDep)
+		env.EventuallyExpectAvgUtilization(v1.ResourceCPU, "<", 0.5)
```
Comment (charliedmcb) [inline test/suites/consolidation/suite_test.go:155]: 0.5 is the % usage?

File: test/pkg/environment/common/setup.go [hunk 4]
```diff
@@ -96,7 +96,12 @@ func (env *Environment) ExpectCleanCluster() {
 		gvk := lo.Must(apiutil.GVKForObject(obj, env.Client.Scheme()))
 		metaList.SetGroupVersionKind(gvk)
 		Expect(env.Client.List(env.Context, metaList, client.Limit(1))).To(Succeed())
-		Expect(metaList.Items).To(HaveLen(0), fmt.Sprintf("expected no %s to exist", gvk.Kind))
+		if len(metaList.Items) > 0 {
+			for _, item := range metaList.Items {
+				// only allow system-surge and default node pools to exist for nap purposes
+				Expect(item.GetName()).To(Or(Equal("system-surge"), Equal("default")), fmt.Sprintf("expected only system-surge and general-purpose %s to exist", gvk.Kind))
```
Comment (charliedmcb) [inline test/pkg/environment/common/setup.go:102]: If its possible to pass in a flag, I'd a way to indicate we are using NAP, or self-hosted. 

It would allow us to have any differences clear within the code, as having clear points like this were we can be like `if NAP then ...` would allow us, and others to easily track any differences.

File: test/pkg/environment/common/expectations.go [hunk 5]
```diff
@@ -97,34 +97,6 @@ func (env *Environment) ExpectCreatedOrUpdated(objects ...client.Object) {
 	}
 }
 
-func (env *Environment) ExpectSettings() (res []v1.EnvVar) {
```
Comment (charliedmcb) [inline test/pkg/environment/common/expectations.go:100]: These are just general cleanups? I'd prefer them to be broken out into a separate PR that I'd be happy to review and approve as to not clutter this one.

File: test/pkg/environment/common/expectations.go [hunk 6]
```diff
@@ -222,30 +194,6 @@ func (env *Environment) ExpectConfigMapDataOverridden(key types.NamespacedName,
 	return true
 }
 
-func (env *Environment) ExpectPodENIEnabled() {
```
Comment (charliedmcb) [inline test/pkg/environment/common/expectations.go:225]: Assuming this is also just general cleanup too. Would prefer it to be moved along with any other general cleanup to a separate PR.

File: test/pkg/environment/common/expectations.go [hunk 7]
```diff
@@ -395,10 +343,18 @@ func (env *Environment) EventuallyExpectUniqueNodeNames(selector labels.Selector
 
 func (env *Environment) eventuallyExpectScaleDown() {
 	GinkgoHelper()
+	var userNodeCount int
 	Eventually(func(g Gomega) {
-		// expect the current node count to be what it was when the test started
-		g.Expect(env.Monitor.NodeCount()).To(Equal(env.StartingNodeCount))
-	}).Should(Succeed(), fmt.Sprintf("expected scale down to %d nodes, had %d", env.StartingNodeCount, env.Monitor.NodeCount()))
+		// Get Nodes that are not part of the systempool. kubernetes.azure.com/mode=user
```
Comment (charliedmcb) [inline test/pkg/environment/common/expectations.go:348]: Assuming this change it to deal with the potential of a NAP cluster getting system surge nodes created?

File: test/suites/consolidation/suite_test.go
Comment (charliedmcb) [inline test/suites/consolidation/suite_test.go:1]: I personally prefer this test to be split into a separate PR just to keep things focus on the new pipeline of "running on nap" in the first PR, and then a secondary PR with a "nap specific e2e test".

File: test/pkg/environment/azure/setup.go
Comment (charliedmcb) [inline test/pkg/environment/azure/setup.go:1]: All this seems to be separate cleanup? Would also prefer it to be pulled out into a cleanup PR.

File: test/pkg/environment/azure/environment.go [hunk 8]
```diff
@@ -51,6 +51,7 @@ func (env *Environment) DefaultNodePool(nodeClass *v1alpha2.AKSNodeClass) *corev
 	nodePool.Spec.Template.Spec.NodeClassRef = &corev1beta1.NodeClassReference{
 		Name: nodeClass.Name,
 	}
+	nodePool.Spec.Weight = lo.ToPtr[int32](1)
```
Comment (charliedmcb) [inline test/pkg/environment/azure/environment.go:54]: Is this needed to make it higher priority than the default NAP pool, or what exactly is this weight impacting?

File: test/pkg/environment/azure/environment.go [hunk 9]
```diff
@@ -84,6 +85,7 @@ func (env *Environment) DefaultNodePool(nodeClass *v1alpha2.AKSNodeClass) *corev
 
 func (env *Environment) ArmNodepool(nodeClass *v1alpha2.AKSNodeClass) *corev1beta1.NodePool {
 	nodePool := env.DefaultNodePool(nodeClass)
+	nodePool.Spec.Weight = lo.ToPtr[int32](1)
```
Comment (charliedmcb) [inline test/pkg/environment/azure/environment.go:88]: Same question on impact here

**Review by charliedmcb — CHANGES_REQUESTED** [review]
Initial review. Didn't get to go over the files within `.github` yet. Will be completing a second review for that.

Overall looking good, but I'd like a couple small adjustments, the ability to signal running the E2E tests for NAP or not via input params, and I think for cleanliness/focus it makes sense to split this into 3 PRs:
- workflow change, and required supporting `Makefile`/`/test` changes
- new consolidation test suite
- general unrelated cleanup to the e2e test files.

File: Makefile-az.mk [hunk 10]
```diff
@@ -3,17 +3,19 @@ ifeq ($(CODESPACES),true)
   AZURE_RESOURCE_GROUP ?= $(CODESPACE_NAME)
   AZURE_ACR_NAME ?= $(subst -,,$(CODESPACE_NAME))
 else
-  AZURE_RESOURCE_GROUP ?= karpenter
+  AZURE_RESOURCE_GROUP ?= karpeeer
```
[the remark this answers is not in the document]
Comment (Bryce-Soghigian) [reply Makefile-az.mk:6]: not on purpose :O

File: Makefile-az.mk [hunk 11]
```diff
@@ -36,6 +38,16 @@ az-mkaks: az-mkacr ## Create test AKS cluster (with --vm-set-type AvailabilitySe
 	az aks get-credentials --name $(AZURE_CLUSTER_NAME) --resource-group $(AZURE_RESOURCE_GROUP) --overwrite-existing
 	skaffold config set default-repo $(AZURE_ACR_NAME).azurecr.io/karpenter
 
+az-mkaks-cilium-nap:
+	az extension add --name aks-preview
```
[the remark this answers is not in the document]
Comment (Bryce-Soghigian) [reply Makefile-az.mk:42]: We need it because there is no way to registere specific versions of the preview cli in the github action from what I could find

File: Makefile-az.mk [hunk 12]
```diff
@@ -36,6 +38,16 @@ az-mkaks: az-mkacr ## Create test AKS cluster (with --vm-set-type AvailabilitySe
...
+	az feature register --namespace "Microsoft.ContainerService" --name "NodeAutoProvisioningPreview"
```
[the remark this answers is not in the document]
Comment (Bryce-Soghigian) [reply Makefile-az.mk:43]: Sub registration could be ran once, but we need the add of the aks-preview extension to be updated to the latest preview cli

File: Makefile-az.mk [hunk 13]
```diff
@@ -36,6 +38,16 @@ az-mkaks: az-mkacr ## Create test AKS cluster (with --vm-set-type AvailabilitySe
...
+	az aks create          --name $(AZURE_CLUSTER_NAME) --resource-group $(AZURE_RESOURCE_GROUP)  \
+		--enable-managed-identity --node-count 3 --generate-ssh-keys -o none --network-dataplane cilium --network-plugin azure --network-plugin-mode overlay \
+		--enable-oidc-issuer --enable-managed-identity --node-provisioning-mode Auto
+
+	az aks get-credentials --name $(AZURE_CLUSTER_NAME) --resource-group $(AZURE_RESOURCE_GROUP) --overwrite-existing
+	kubectl taint nodes CriticalAddonsOnly=true:NoSchedule --all
```
[the remark this answers is not in the document]
Comment (Bryce-Soghigian) [reply Makefile-az.mk:49]: They should ALWAYS taint the systempool. NAP should create all of the userpool nodes imo.

File: test/pkg/environment/common/expectations.go (hunk 7 above)
Comment (Bryce-Soghigian) [reply test/pkg/environment/common/expectations.go:348]: yep exactly

File: test/suites/consolidation/suite_test.go
Comment (Bryce-Soghigian) [reply test/suites/consolidation/suite_test.go:1]: Sure can split it out

File: test/pkg/environment/azure/setup.go
Comment (Bryce-Soghigian) [reply test/pkg/environment/azure/setup.go:1]: sure can split up

File: Makefile-az.mk (hunk 11 above)
[the remark this answers is not in the document]
Comment (charliedmcb) [reply Makefile-az.mk:42]: I'd still be wary on adding it to this make command. Separating it out as an `az-apply-nap-prerequirements` make command along with the feature registration as I mention in the below comment I think would be good. That way in the action we can call `az-apply-nap-prerequirements`, but anyone doing local work, or such can avoid having the `prerequirements` side effects.

File: test/suites/consolidation/suite_test.go (hunk 3 above)
Comment (Bryce-Soghigian) [reply test/suites/consolidation/suite_test.go:155]: yep

File: Makefile-az.mk (hunk 13 above)
[the remark this answers is not in the document]
Comment (charliedmcb) [reply Makefile-az.mk:49]: Wonder if we should add something about it to the docs than?
- https://learn.microsoft.com/en-us/azure/aks/node-autoprovision?tabs=azure-cli

So far tainting has been something we've done within the `e2e` test workflow themselves not the make commands.

However, if it is a best practice we want to recommend than I could see adding it to the make command, and/or documentation.

[the remark this answers is not in the document]
Comment (Bryce-Soghigian) [reply Makefile-az.mk:49]: I agree we should! Maybe referencing the existing systempool documentation makes sense and explaining a bit more about the role of the system-surge pool

File: test/pkg/environment/azure/environment.go (hunk 8 above)
Comment (Bryce-Soghigian) [reply test/pkg/environment/azure/environment.go:54]: # Priority given to the NodePool when the scheduler considers which NodePool
  # to select. Higher weights indicate higher priority when comparing NodePools.
  # Specifying no weight is equivalent to specifying a weight of 0.
  weight: 10
  
  
 we want it to be higher weighted than the system surge and other pools since we are testing specific nodepool functionality. I could set this to 10 instead

File: .github/workflows/e2e-matrix-trigger-nap.yaml [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+  resolve:
+    if: github.event_name != 'workflow_run' || github.event.workflow_run.conclusion == 'success'
+    uses: ./.github/workflows/resolve-args.yaml
+  e2e-matrix:
+    permissions:
+      contents: read
+      id-token: write # This is required for requesting the JWT
+      statuses: write # ./.github/actions/commit-status/*          
+    needs: [resolve]
+    uses: ./.github/workflows/e2e-matrix-nap.yaml
```
Comment (charliedmcb) [inline .github/workflows/e2e-matrix-trigger-nap.yaml:27]: This file is so similar to the non-nap one that I think we could pass along a param on if its running in nap, or not

File: .github/workflows/e2e-matrix-nap.yaml [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+        run: |
+          E2E_HASH="$RANDOM$RANDOM"
+          echo "Using e2e hash \"$E2E_HASH\""
+          echo "E2E_HASH=$E2E_HASH" >> "$GITHUB_OUTPUT"
+  e2e:
+    needs: [initialize-generative-params]
+    strategy:
+      fail-fast: false
+      matrix:
+        suite: [Nonbehavioral, GPU, Integration, NodeClaim, Consolidation]
```
Comment (charliedmcb) [inline .github/workflows/e2e-matrix-nap.yaml:46]: This, and using the `e2e-nap` is really the only differences. I think its similar enough we could pass along a param on if its running in nap, or not

Comment (charliedmcb) [reply .github/workflows/e2e-matrix-nap.yaml:46]: At this point we could run one of two matrix sets based on nap or not. At least I think this is reasonable.

File: .github/workflows/e2e-nap.yaml [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+        run: |
+          # Creating jitter so that we can stagger cluster creation to avoid throttling
+          sleep $(( RANDOM % 60 + 1 ))
+      - name: generate cluster name
+        run: |
+          CLUSTER_NAME=mc
+          echo "Using cluster name \"$CLUSTER_NAME\""
+          echo "CLUSTER_NAME=$CLUSTER_NAME" >> "$GITHUB_ENV"
+      - name: create aks cluster '${{ env.CLUSTER_NAME }}'
+        uses: ./.github/actions/e2e/create-nap-cluster
```
Comment (charliedmcb) [inline .github/workflows/e2e-nap.yaml:89]: This and cleanup at the only difference. I wonder if we could again consolidate into just one `e2e.yaml` with the ability to trigger either a nap cluster creation or regular, based on a input flag.

File: .github/actions/e2e/create-nap-cluster/action.yaml
Comment (charliedmcb) [inline .github/actions/e2e/create-nap-cluster/action.yaml:1]: Lot of duplication, but I think it makes sense to have a separate action for nap like this.

File: .github/actions/e2e/cleanup-nap/action.yaml [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+      with:
+        client-id: ${{ inputs.client-id }}
+        tenant-id: ${{ inputs.tenant-id }}
+        subscription-id: ${{ inputs.subscription-id }}
+    - name: az set sub
+      shell: bash
+      run: az account set --subscription ${{ inputs.subscription-id }}
+    - name: delete cluster ${{ inputs.cluster_name }}
+      shell: bash
+      run: az aks delete --name ${{ inputs.cluster_name }} --resource-group ${{ inputs.resource_group }} --yes --no-wait
```
Comment (charliedmcb) [inline .github/actions/e2e/cleanup-nap/action.yaml:45]: Just doesn't have the delete `acr`? set. Could simply remove that from the general cleanup since resource group delete should be enough. I think that's worth removing the need for an almost duplicate file.

**Review by charliedmcb — CHANGES_REQUESTED** [review]
Little mixed on how much to duplicate the workflow files, but I think at least a few of them can definitely be merged with the regular flow:
- approval-comment-nap.yaml (merge with regular)
- cleanup-nap/action.yaml (merge with regular)

I'd try and merge based on some form of flag:
- e2e-matrix-trigger-nap.yaml
- e2e-matrix-nap.yaml

I think that e2e-nap.yaml might be able to be consolidated, but I'm wondering if its right to as it somewhat seems right that a "nap e2e test is different". I think at least through the point of we are selecting the matrix of tests to run, and how to run them seems reasonable.

I think keeping a separate create action for nap does make sense though.

Not certain on all the functionality available for getting some sort of flag for this, and if it'd be too messy in the code, but those are my general thoughts on reducing duplication for now.

File: test/pkg/environment/azure/setup.go
Comment (Bryce-Soghigian) [reply test/pkg/environment/azure/setup.go:1]: https://github.com/Azure/karpenter-provider-azure/pull/240

File: test/pkg/environment/common/expectations.go (hunk 6 above)
Comment (Bryce-Soghigian) [reply test/pkg/environment/common/expectations.go:225]: https://github.com/Azure/karpenter-provider-azure/pull/240

**Comment by Bryce-Soghigian** [comment]
cleanup prs separated out: 
https://github.com/Azure/karpenter-provider-azure/pull/241
https://github.com/Azure/karpenter-provider-azure/pull/240

File: test/suites/consolidation/suite_test.go (hunk 1 above)
Comment (Bryce-Soghigian) [reply test/suites/consolidation/suite_test.go:58]: Resolving since we removed consolidation

**Comment by Bryce-Soghigian** [comment]
test ping @Azure/aks-karpenter
